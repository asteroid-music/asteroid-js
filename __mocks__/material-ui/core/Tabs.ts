export default "Material UI Tabs";