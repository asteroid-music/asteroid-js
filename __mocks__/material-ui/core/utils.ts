function createSvgIcon(path,displayName) {
    return "Memo("+displayName+"Icon)";
}

export { createSvgIcon };
