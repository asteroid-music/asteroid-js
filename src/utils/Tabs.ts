/**
 * Interface describing an object with the same elements as SubTabObject
 */
interface SubTabObjectLike {
    /** The name of the tab */
    name: string;

    /** A string array containing names of subtabs */
    component: React.ComponentType<any>;
}

/**
 * Interface describing an object with the same elements as TabObject
 */
interface TabObjectLike {
    /** The name of the tab */
    name: string;

    /** A subtab array containing names of subtabs */
    subtabs: SubTabObjectLike[];
}

class SubTabObject {
    /** The name of the subtab */
    name: string;

    /** The component of the subtab */
    component: React.ComponentType<any>;

    /**
     * Constructor method
     *
     * @param {object} rawJson: a raw json representing the object
     */
    constructor(rawJson: SubTabObjectLike) {
        this.name = rawJson.name;
        this.component = rawJson.component;
    }

}

/**
 * Class describing named tab, along with strings representing all the subtabs
 * it contains
 */
class TabObject {
    /** The name of the tab */
    name: string;

    /** A string array containing names of subtabs */
    subtabs: SubTabObject[];

    /**
     * Constructor method
     *
     * @param {object} rawJson: a raw json representing the object
     */
    constructor(rawJson: TabObjectLike) {
        this.name = rawJson.name;
        this.subtabs = rawJson.subtabs.map((item: SubTabObjectLike) => {
            return new SubTabObject(item);
        });
    }

    /**
     * Gets a list of names of subtabs in the tab
     *
     * @returns {string[]} tablist: the name of each subtab
     */
    nameList() {
       return this.subtabs.map((subtab: SubTabObject) => {
            return subtab.name;
       });
    }

    /**
     * Checks if the subtab is present in the tab
     *
     * @param {string} subtab: the name of the subtab to check for
     *
     * @returns {boolean} isin: true if subtab is in the tab; false otherwise
     */
    includes(subtab:string) {
        return this.nameList().includes(subtab);
    }

    /**
     * Returns the subtab specified by the given subtab name,
     *      or 'null' if none exists
     *
     * @param {string} subtab: the name of the subtab to get
     *
     * @returns {SubTabObject | null} subtab: the specified subtab if exists;
     *      'null' otherwise
     */
    get(subtab:string) {
        let objIndex: number = this.nameList().indexOf(subtab);
        if (objIndex === -1) {
            return null;
        } else {
            return this.subtabs[objIndex];
        }
    }
}

/**
 * Class describing an array of named tabs
 */
class TabArray {
    /** Encapsulated array */
    tabs: TabObject[];

    /**
     * Constructor method
     *
     * @param {object} rawJson: a raw json representing the object
     */
    constructor(rawJson: TabObjectLike[]) {
        this.tabs = rawJson.map((item: TabObjectLike) => {
            return new TabObject(item);
        });
    }

    /**
     * Gets a list of names of tabs in the array
     *
     * @returns {string[]} tablist: the name of each tab
     */
    nameList() {
       return this.tabs.map((tab: TabObject) => {
            return tab.name;
       });
    }

    /**
     * Checks if the tab is present in the array
     *
     * @param {string} tab: the name of the tab to check for
     *
     * @returns {boolean} isin: true if tab is in the tab array;
     *      false otherwise
     */
    includes(tab:string) {
        return this.nameList().includes(tab);
    }

    /**
     * Returns the tab specified by the given tabname, or 'null' if none exists
     *
     * @param {string} tab: the name of the tab to get
     *
     * @returns {TabObject | null} tab: the specified tab if exists;
     *      'null' otherwise
     */
    get(tab:string) {
        let objIndex: number = this.nameList().indexOf(tab);
        if (objIndex === -1) {
            return null;
        } else {
            return this.tabs[objIndex];
        }
    }
}

export { TabObjectLike, TabObject, TabArray }
