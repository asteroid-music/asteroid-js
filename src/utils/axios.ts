import axios from 'axios';

//console.log(asteroidConfig);

declare var __config__: {
    "axios": {
        "baseURL": string,
        "timeout": number
    }
};

let instance = axios.create(__config__.axios || {});

export default instance;
