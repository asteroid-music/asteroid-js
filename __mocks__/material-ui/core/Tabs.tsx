import React from 'react';
import { serialize as serialise } from 'jest-snapshot/build/utils';
//I know, UK vs US spelling...

const Tabs = props => <div
>
    Material UI Tabs with props:
    {
        serialise(props)
    }
    {props.onChange ? props.children.map((child)=>{
        return React.cloneElement(child,{onChange: (event,value)=>{props.onChange(event,value);}});
    }) : props.children}
</div>

export default Tabs;