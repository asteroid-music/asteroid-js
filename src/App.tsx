//Import from external module 'react'
import React from 'react';

//Import from external module '@material-ui/core'
import Box from '@material-ui/core/Box';

//Import from external module 'react-router-dom'
import { HashRouter, Switch, Route } from 'react-router-dom';

//Import from local src/
import TopBar from 'parts/TopBar';
import { TabArray } from 'utils/Tabs';
import NoOpenTab from 'views/NoOpenTab';
import NoMatchedTab from 'views/NoMatchedTab';

/**
 * Utility interface describing props for the main React <App>
 */
interface AppProps {
    /** A TabArray object representing the accessible tabs */
    tabs: TabArray;
}

function App(props: AppProps) {
    if (props.tabs && props.tabs instanceof TabArray) {
        return (
            <HashRouter>
                <Box width="100%">
                    <TopBar
                        tabs={props.tabs}
                    />
                    <Switch>
                        <Route exact path="/">
                            <NoOpenTab />
                        </Route>
                        {props.tabs.nameList().map((tabName)=>{
                            return <Route path={"/"+tabName} key={tabName}>
                                <Switch>
                                    {props.tabs.get(tabName)?.nameList().map((subTabName)=>{
                                        return <Route
                                            key={tabName+"/"+subTabName}
                                            path={"/"+tabName+"/"+subTabName}
                                            component={props.tabs.get(tabName).get(subTabName)?.component}
                                        />
                                    })}
                                    <Route path="*">
                                        <NoMatchedTab />
                                    </Route>
                                </Switch>
                            </Route>
                        })}
                        <Route path="*">
                            <NoMatchedTab />
                        </Route>
                    </Switch>
                </Box>
            </HashRouter>
        );
    } else {
        throw Error("Unable to render the <App> component: invalid props. This is likely an issue with the server and not your browser.");
    }
}

export default App;
