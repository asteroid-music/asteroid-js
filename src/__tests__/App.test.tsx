import React from 'react';
import renderer from 'react-test-renderer';
import App from 'App';

/**
 * App.tsx imports:
 *  - React
 *  - Box
 *  - react-router-dom
 *  - TopBar
 *  - TabArray
 *  - NoOpenTab
 *  - NoMatchedTab
 * 
 * The following should be mocked:
 *  - Box
 *  - TopBar
 *  - TabArray
 *  - NoOpenTab
 *  - NoMatchedTab
 */

jest.mock("parts/TopBar", ()=>"TopBar");
jest.mock("views/NoOpenTab", ()=>"NoOpenTab");
jest.mock("views/NoMatchedTab", ()=>"NoMatchedTab");
jest.mock("@material-ui/core/Box", ()=>"Material UI Box");

import { TabArray } from 'utils/Tabs';

const tabs = new TabArray([
    {
        name:"Testing Tab 1",
        subtabs: [
            {
                name: "Testing Subtab 1",
                component: (props)=><div>TestingSubtab1Component</div>
            },
            {
                name: "Testing Subtab 2",
                component: (props)=><div>TestingSubtab2Component</div>
            }
        ]
    },
    {
        name:"Testing Tab 2",
        subtabs: [
            {
                name: "Testing Subtab 3",
                component: (props)=><div>TestingSubtab3Component</div>
            },
            {
                name: "Testing Subtab 3",
                component: (props)=><div>TestingSubtab4Component</div>
            }
        ]
    },
]);

describe('<App />',()=>{
    //Basic rendering with minimal props
    describe('basic rendering',()=>{
        /**
         * The App props are:
         *  tabs: TabArray
         * 
         * Each of these should be tested both with valid and invalid values
         */
        it("renders correctly",()=>{
            expect(renderer
                .create(<App tabs={tabs}/>)
                .toJSON()
            ).toMatchSnapshot();
        });
    });

    //Testing out location functionalities
    describe('location-based rendering',()=>{
        /**
         * Different behaviours when:
         *  - at "/"
         *  - at a location included in tabs
         *  - at a location not included in tabs
         */
    });
});