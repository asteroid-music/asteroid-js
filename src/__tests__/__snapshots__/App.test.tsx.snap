// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> basic rendering renders correctly 1`] = `
<Material UI Box
  width="100%"
>
  <TopBar
    tabs={
      TabArray {
        "tabs": Array [
          TabObject {
            "name": "Testing Tab 1",
            "subtabs": Array [
              SubTabObject {
                "component": [Function],
                "name": "Testing Subtab 1",
              },
              SubTabObject {
                "component": [Function],
                "name": "Testing Subtab 2",
              },
            ],
          },
          TabObject {
            "name": "Testing Tab 2",
            "subtabs": Array [
              SubTabObject {
                "component": [Function],
                "name": "Testing Subtab 3",
              },
              SubTabObject {
                "component": [Function],
                "name": "Testing Subtab 3",
              },
            ],
          },
        ],
      }
    }
  />
  <NoOpenTab />
</Material UI Box>
`;
